text/0000-native-class-roadmap.md

```diff
@@ -0,0 +1,265 @@
+community for pursuing new language features and libraries (e.g. mixins)
+- **It should be possible to adopt incrementally.** This will be one of the largest
+changes the community has ever gone through, on the same level if not larger than
+dropping Views and adopting Components. EmberObject is pervasive - it's been
+used for every construct imaginable in Ember apps and addons due to its
+versatility and simplicity. Ideally users will be able to make incremental
+changes toward pure native classes by first rewriting their existing classes
+using native class syntax (as is already possible) and then dropping EmberObject
+altogether.
+- **We should consider the long tail of adoption, particularly for addons.** There
+will be a long period where the community will need to support older versions of
+Ember at the same time as newer versions. Ideally it should be possible to
+support both the new and old model at once without separate code paths, or the
+overlap period should be long enough that addons can drop EmberObject without
+breaking Ember versions that are still part of their support matrix.
+- **If possible, conversion should be automate-able.** This is not a hard
+constraint, since it limits us to having very similar behavior to the existing
+object model. However, if it is possible to create scripts that speed up
+adoption it will help the community tremendously, so if we are at an impasse on
+particular behaviors or implementations, we should consider whether one would be
+easier to automate towards.
+
+## Detailed design
+
+As mentioned above, one of the main reasons this conversion is going to be a
+difficult and long process is the fact that both user code and Ember code are
+very intertwined with the existing object model. Specifically:
+
+- They rely on the behavior of static methods such as `create`, `extend`, and
+  `destroy`.
+- They rely on the behavior of EmberObject's constructor
+- They rely on a non-trivial amount of functionality provided via inheritance
+- They rely on features of EmberObject's inheritance model that do not exist in
+native Javascript (Mixins, concatenated/merged properties, nuances of observers
+and listeners, etc.)
+
+The fact that these are so intertwined makes it very difficult (if not
+impossible) for us to rewrite the core classes that exist in Ember without
+breaking user land code.
+
+One strategy we can implement to prevent this from happening again in the future
+is to make the details of classes opaque to Ember. Rather than providing base
+classes that can be extended by users to add functionality and implement their
+applications, we can rewrite Ember to treat user classes as delegates. In other
+words, Ember will make no assumptions about the behavior of a class, other than
+that it fulfills a certain interface.
+
+For instance, a Component can be any class which receives an `args` hash and
+optionally implements any of the component lifecycle hooks:
```

Comment:
Non important, curious question - Would this in effect reduce lookup time for properties/non existent properties, especially for things are are high up the prototype chain?

Reply:
In theory yes, though I wouldn't count on it being that huge of a win. In aggregate though, who knows 😄